tools/pubsys/src/repo/fetch_variant/mod.rs

```diff
@@ -0,0 +1,265 @@
+//! The fetch_variant module owns the 'fetch-variant' subcommand and provides methods for fetching
+//! a given variant and download its image targets.
+
+use crate::repo::{error as repo_error, repo_urls};
+use crate::{repo, Args};
+use clap::Parser;
+use log::{info, trace};
+use pubsys_config::InfraConfig;
+use snafu::{OptionExt, ResultExt};
+use std::path::{Path, PathBuf};
+use tempfile::tempdir;
+use tough::{Prefix, Repository, RepositoryLoader, TargetName};
+use url::Url;
+use walkdir::WalkDir;
+
+use buildsys::manifest::{ImageFormat, ManifestInfo, PartitionPlan};
+
+/// fetching and downdloaing the image targets of a given variant
+#[derive(Debug, Parser)]
+pub(crate) struct FetchVariantArgs {
+    #[arg(long)]
+    /// Use this named repo infrastructure from Infra.toml
+    repo: String,
+
+    #[arg(long)]
+    /// The architecture of the repo being validated
+    arch: String,
+
+    #[arg(long)]
+    /// The variant of the repo being validated
+    variant: String,
+
+    #[arg(long)]
+    /// Path to root.json for this repo
+    root_role_path: PathBuf,
+
+    #[arg(long)]
+    /// Where to store the downloaded img files
+    outdir: PathBuf,
+
+    #[arg(long)]
+    /// The variant name without extension
+    filename_prefix: String,
+
+    #[arg(long)]
+    /// The root directory of the buildsys
+    variant_manifest: PathBuf,
+}
+
+/// Download targets
+async fn handle_download(
+    repository: &Repository,
+    outdir: &Path,
+    raw_names: &[String],
+) -> Result<(), Error> {
+    let target_names: Result<Vec<TargetName>, Error> = raw_names
+        .iter()
+        .map(|s| TargetName::new(s).context(error::InvalidTargetNameSnafu))
+        .collect();
+    let target_names = target_names?;
+
+    if target_names.is_empty() {
+        return error::MissingTargetNamesSnafu.fail();
+    };
+
+    let tempdir = tempdir().context(error::CreateTempDirSnafu)?;
+    let tempdir_path = tempdir.path();
+    let download_target = |name: TargetName| async move {
+        info!("\t-> {}", name.raw());
+        repository
+            .save_target(&name, tempdir_path, Prefix::None)
+            .await
+            .context(error::SaveTargetSnafu)
+    };
+
+    info!("Downloading targets to {tempdir_path:?}");
+    for target in target_names {
+        download_target(target).await?;
+    }
+
+    info!("Cleaning up {outdir:?}");
+    tokio::fs::remove_dir_all(outdir)
+        .await
+        .context(error::CleanDirSnafu { path: outdir })?;
+
+    info!("Moving targets to {outdir:?}");
+    tokio::fs::create_dir_all(outdir)
+        .await
+        .context(error::CreateDirSnafu { path: outdir })?;
+    for tempdir_direntry in WalkDir::new(outdir) {
+        let tempdir_target = tempdir_direntry.context(error::DirectoryWalkSnafu)?;
+        tokio::fs::copy(tempdir_target.path(), outdir)
+            .await
+            .context(error::CopyTargetSnafu)?;
+    }
```

Comment:
Something isn't quite right here:
```shell
Failed to fetch variant: Failed to copy target: the source path is neither a regular file nor a symlink to a regular file
```

Reply:
This is actually just a call to `asyncify(|| std::fs::copy(from, to)).await` and `std::fs::copy` throws that error

Reply:
Since we're already walking through the tempdir, I'd assume this means we have more directories inside it that we need to recursively copy.

Reply:
I was silly. We don't need to walkdir because we already know what these filenames are.